File: Lernwerk/Views/Present/PresentationImport.swift
import PDFKit
import UIKit
import UniformTypeIdentifiers

/// Turns a PowerPoint file or a PDF into a presentation in the store. A PDF becomes one picture slide per page,
/// with the page text kept for the AI.
enum PresentationImport {
    static let pptxType = UTType("org.openxmlformats.presentationml.presentation") ?? UTType(filenameExtension: "pptx") ?? .data
    static let contentTypes: [UTType] = [pptxType, .pdf]

    enum ImportError: LocalizedError {
        case unreadable
        case legacyPowerPoint
        case unreadablePDF

        var errorDescription: String? {
            switch self {
            case .unreadable: return "Die Datei lässt sich nicht öffnen."
            case .legacyPowerPoint: return "Alte PowerPoint-Dateien (.ppt) gehen nicht. Speichere sie in PowerPoint oder Keynote als .pptx und importiere sie dann."
            case .unreadablePDF: return "Die PDF lässt sich nicht öffnen."
            }
        }
    }

    struct Result {
        var presentation: Presentation
        /// Objects that could not be taken over (charts, SmartArt, vector pictures).
        var skipped: Int
    }

    @MainActor
    static func run(_ url: URL, store: PresentationStore) async throws -> Result {
        let scoped = url.startAccessingSecurityScopedResource()
        defer { if scoped { url.stopAccessingSecurityScopedResource() } }
        guard let data = try? Data(contentsOf: url) else { throw ImportError.unreadable }
        let title = url.deletingPathExtension().lastPathComponent
        let ext = url.pathExtension.lowercased()
        if ext == "ppt" { throw ImportError.legacyPowerPoint }

        let result: Result
        if ext == "pdf" || data.starts(with: Data("%PDF".utf8)) {
            result = try pdf(data, title: title, store: store)
        } else {
            let imported: PptxReader.Imported
            do {
                imported = try await Task.detached(priority: .userInitiated) { try PptxReader.read(data, title: title) }.value
            } catch {
                throw ImportError.unreadable
            }
            for (name, bytes) in imported.media { try store.saveMedia(bytes, named: name) }
            result = Result(presentation: imported.presentation, skipped: imported.skipped)
        }
        store.add(result.presentation)
        return result
    }

    @MainActor
    private static func pdf(_ data: Data, title: String, store: PresentationStore) throws -> Result {
        guard let document = PDFDocument(data: data), document.pageCount > 0 else { throw ImportError.unreadablePDF }
        let texts = PDFMaterialReader.pageTexts(of: document)
        var slides: [Slide] = []
        for index in 0..<document.pageCount {
            guard let page = document.page(at: index) else { continue }
            let image = page.thumbnail(of: CGSize(width: 1920, height: 1920), for: .cropBox)
            guard let jpeg = image.jpegData(compressionQuality: 0.85) else { continue }
            let name = try store.saveMedia(jpeg, fileExtension: "jpg")
            let frame = SlideGeometry.fit(width: image.size.width, height: image.size.height, into: (0, 0, SlideSize.width, SlideSize.height))
            slides.append(Slide(
                elements: [SlideElement(kind: .image, x: frame.x, y: frame.y, width: frame.width, height: frame.height, image: name)],
                extractedText: index < texts.count ? texts[index] : "",
                background: "#FFFFFF"
            ))
        }
        guard !slides.isEmpty else { throw ImportError.unreadablePDF }
        return Result(presentation: Presentation(title: title, themeId: SlideTheme.paper.id, slides: slides), skipped: 0)
    }
}

File: android/app/src/main/java/de/maxifrz/lernwerk/present/PresentationImport.kt
package de.maxifrz.lernwerk.present

import android.content.Context
import android.net.Uri
import android.provider.OpenableColumns
import de.maxifrz.lernwerk.data.ImageCompressor
import de.maxifrz.lernwerk.data.PresentationStore
import de.maxifrz.lernwerk.pdf.PdfPages
import de.maxifrz.lernwerk.pdf.PdfText
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.File
import java.io.IOException

/** Turns a PowerPoint or PDF file into an editable presentation in the store. */
object PresentationImport {
    class Result(val presentation: Presentation, val skipped: Int)

    suspend fun import(context: Context, store: PresentationStore, uri: Uri): Result = withContext(Dispatchers.IO) {
        val resolver = context.contentResolver
        val name = resolver.query(uri, arrayOf(OpenableColumns.DISPLAY_NAME), null, null, null)?.use { cursor ->
            if (cursor.moveToFirst()) cursor.getString(0) else null
        } ?: uri.lastPathSegment?.substringAfterLast('/') ?: "Präsentation"
        val title = name.substringBeforeLast('.').ifBlank { "Präsentation" }
        val bytes = resolver.openInputStream(uri)?.use { it.readBytes() } ?: throw IOException("Die Datei lässt sich nicht öffnen.")
        val type = resolver.getType(uri).orEmpty()
        val isPdf = type == "application/pdf" || name.endsWith(".pdf", ignoreCase = true) ||
            (bytes.size > 4 && String(bytes, 0, 4, Charsets.ISO_8859_1) == "%PDF")
        val result = if (isPdf) pdf(context, store, bytes, title) else pptx(store, bytes, title)
        withContext(Dispatchers.Main) { store.add(result.presentation) }
        result
    }

    private suspend fun pptx(store: PresentationStore, bytes: ByteArray, title: String): Result {
        val imported = runCatching { PptxReader.read(bytes, title) }
            .getOrElse { throw IOException("Das ist keine lesbare PowerPoint-Datei (.pptx). Ältere .ppt-Dateien vorher in PowerPoint als .pptx speichern.") }
        imported.media.forEach { (mediaName, data) -> store.saveMedia(mediaName, data) }
        return Result(imported.presentation, imported.skipped)
    }

    /** Every page becomes a picture slide; its text layer goes along for the AI. */
    private suspend fun pdf(context: Context, store: PresentationStore, bytes: ByteArray, title: String): Result {
        val file = File.createTempFile("import", ".pdf", context.cacheDir)
        try {
            file.writeBytes(bytes)
            val texts = PdfText.pageTexts(file).orEmpty()
            val pages = PdfPages.open(file) ?: throw IOException("Das PDF lässt sich nicht öffnen.")
            val slides = pages.use { document ->
                (0 until document.pageCount).mapNotNull { index ->
                    val bitmap = document.render(index, 1920) ?: return@mapNotNull null
                    val media = store.saveMedia(ImageCompressor.encode(bitmap, 85), "jpg")
                    val frame = SlideGeometry.fit(bitmap.width, bitmap.height, SlideGeometry.RectBox(0f, 0f, SlideSize.WIDTH, SlideSize.HEIGHT))
                    Slide(
                        elements = listOf(SlideElement(kind = ElementKind.IMAGE, x = frame.x, y = frame.y, width = frame.width, height = frame.height, image = media)),
                        extractedText = texts.getOrNull(index).orEmpty(),
                        background = "#FFFFFF",
                    )
                }
            }
            if (slides.isEmpty()) throw IOException("Das PDF hat keine lesbaren Seiten.")
            return Result(Presentation(title = title, themeId = SlideTheme.PAPER.id, slides = slides), 0)
        } finally {
            file.delete()
        }
    }
}
